django/forms/models.py

```diff
@@ -448,19 +404,46 @@ def validate_unique(self):
 
     def save(self, commit=True):
         """
-        Saves this ``form``'s cleaned_data into model instance
-        ``self.instance``.
-
-        If commit=True, then the changes to ``instance`` will be saved to the
-        database. Returns ``instance``.
+        Save this form's self.instance object if commit=True. Otherwise, add
+        a save_m2m() method to the form which can be called after the instance
+        is saved manually at a later time. Return the model instance.
         """
-        if self.instance.pk is None:
-            fail_message = 'created'
+        instance = self.instance
+        fields = self._meta.fields
+        opts = instance._meta
+        if self.errors:
+            raise ValueError(
+                "The %s could not be %s because the data didn't validate." % (
+                    opts.object_name,
+                    'created' if instance._state.adding else 'changed',
+                )
+            )
+
+        # Wrap up the saving of m2m data as a function.
+        def save_m2m():
+            exclude = self._meta.exclude
+            cleaned_data = self.cleaned_data
+            # Note that for historical reasons we want to include also
+            # virtual_fields here. (GenericRelation was previously a fake
+            # m2m field).
+            for f in chain(opts.many_to_many, opts.virtual_fields):
+                if not hasattr(f, 'save_form_data'):
+                    continue
+                if fields and f.name not in fields:
+                    continue
+                if exclude and f.name in exclude:
+                    continue
+                if f.name in cleaned_data:
+                    f.save_form_data(instance, cleaned_data[f.name])
```

Comment:
While you're around, do you think it would make sense to move this inner function to a `_save_m2m` method?